### ISSUE TITLE
describe further goals in issues or milestones

### ISSUE THREAD

**Issue by titaniumbones** [ubolonton/emacs-tree-sitter#17]
I['m entirely out of my depth here, as I have only the most minimal understanding of tree-sitter and the approach you're taking here. But I wonder if you could add some issues to track (even aspirational) goals for hte project beyond syntax highlighting. For instance, in the readme you talk about these features:

- More flexible code folding.
- Structural editing (like Paredit, or even better) for non-Lisp code.
- More informative indexing for imenu.

Maybe they don't actually belong in this package, but in some extensions (esp since people have their own idiosyncratic preferences for things like lispy/paredit and code folding); but it would be pretty nice to see a bit more fleshed out ideas of how these features would work, and what's in principle possible now, etc etc.  

Really excited to see this project -- fast and perfect syntax highlighting on its own will be a huge win, and if structural editing in all languages comes around too, it will be fantastic.

**Comment by rirze** [comment]
I was thinking about the structural editing aspect a few days ago-- I believe tree-sitter could be a nice backbone for a revised version of `expand-region`, for example. If not, then I could also see a more general `paredit`-like package.  

This could be a very popular use case for tree-sitter, if getting the attention of the community were a priority.

**Comment by ubolonton** [comment]
I'm working on publishing pre-compiled modules and grammars, to enable people to immediately work with the Lisp code without having to touch/compile the Rust part (#2).

I'll create more issues to discuss potential ideas later this week.

**Comment by ubolonton** [comment]
I created some more issues.

I intend to write a guide on using the APIs soon, so that it becomes more approachable to other lib writers.

**Comment by sogaiu** [comment]
> I was thinking about the structural editing aspect a few days ago-- I believe tree-sitter could be a nice backbone for a revised version of `expand-region`, for example. If not, then I could also see a more general `paredit`-like package.

It's not `expand-region`, but I experimented with selecting top-levels here: https://github.com/sogaiu/ab/blob/master/ab-tree-sitter.el#L74-L88

This seems to work in a number of languages for which tree-sitter is available.  I've tried c, clojure, python, and rust with some success.

On a somewhat related note I'm interested in seeing how code folding might work.

Does anyone have a working example?
